lib/Web/Server.hs

```diff
@@ -29,18 +28,142 @@ import Web.API
+  res@DFREntryNotFound{} ->
+    throwError $ addApplicationJSONHeader err404 { errBody = cofferServerErrorToBytes (pretty res) 400 }
+  res@DFRFieldNotFound{} ->
+    throwError $ addApplicationJSONHeader err404 { errBody = cofferServerErrorToBytes (pretty res) 401 }
+  where
+    pretty = resultToText buildDeleteFieldResult
+
+throwCopyOrRenameError :: Bool -> CopyResult -> Handler a
+throwCopyOrRenameError rename = \case
+  CPRSuccess{} -> throwError err500 { errBody = "Expected " <> cmdName <> " error, got CPRSuccess." }
+  res@CPRPathNotFound{} ->
+    throwError $ addApplicationJSONHeader err404 { errBody = cofferServerErrorToBytes (prettySingleMessage res) 500 }
+  res@CPRMissingEntryName{} ->
+    throwError $ addApplicationJSONHeader err400 { errBody = cofferServerErrorToBytes (prettySingleMessage res) 501 }
+  res@CPRSamePath{} ->
+    throwError $ addApplicationJSONHeader err409 { errBody = cofferServerErrorToBytes (prettySingleMessage res) 502 }
+  res@(CPRCreateErrors createErrors) -> do
+    let messages = buildCopyOrRenameResult rename WebAPI res <&> pretty
+    let codes = createErrors <&> createErrorToCode . snd
+    let serverErrors = zipWith CofferServerError messages codes
+    throwError $ addApplicationJSONHeader err409 { errBody = encode serverErrors }
+  where
+    cmdName = if rename then "rename" else "copy"
+
+    prettySingleMessage :: CopyResult -> Text
+    prettySingleMessage res =
+      buildCopyOrRenameResult rename WebAPI res
+        & unlinesF
+        & pretty
+
+    createErrorToCode :: CreateError -> Int
+    createErrorToCode = \case
+      CEParentDirectoryIsEntry{} -> 503
+      CEDestinationIsDirectory{} -> 504
+      CEEntryAlreadyExists{} -> 505
+
+throwCopyError :: CopyResult -> Handler a
+throwCopyError = throwCopyOrRenameError False
 
-mount :: Text
-mount = "secret"
+throwRenameError :: RenameResult -> Handler a
+throwRenameError = throwCopyOrRenameError True
+
+throwDeleteError :: DeleteResult -> Handler a
+throwDeleteError = \case
+  DRSuccess{} -> throwError err500 { errBody = "Expected delete error, got DRSuccess." }
+  res@DRPathNotFound{} ->
+    throwError $ addApplicationJSONHeader err404 { errBody = cofferServerErrorToBytes (pretty res) 600 }
+  res@DRDirectoryFound{} ->
+    throwError $ addApplicationJSONHeader err400 { errBody = cofferServerErrorToBytes (pretty res) 601 }
+  where
+    pretty = resultToText buildDeleteResult
+
+throwTagError :: TagResult -> Handler a
+throwTagError = \case
+  TRSuccess{} -> throwError err500 { errBody = "Expected tag error, got TRSuccess." }
+  res@TREntryNotFound{} ->
+    throwError $ addApplicationJSONHeader err404 { errBody = cofferServerErrorToBytes (pretty res) 700 }
+  res@TRTagNotFound{} ->
+    throwError $ addApplicationJSONHeader err404 { errBody = cofferServerErrorToBytes (pretty res) 701 }
+  res@TRDuplicateTag{} ->
+    throwError $ addApplicationJSONHeader err409 { errBody = cofferServerErrorToBytes (pretty res) 702 }
```

Comment:
We should encapsulate this pattern in a single function and use it everywhere, e.g.:

```hs
throwCofferServerError :: ServerError -> Int -> Text -> Handler a
```

This single function would be responsible for creating the json body and adding the content-type header.

Reply:
And a more generic version of it, to use in `CPRCreateErrors`
```hs
throwCofferServerErrors :: ServerError -> [(Int, Text)] -> Handler a
```

Reply:
Done